pkg/controller/admissionchecks/provisioning/controller.go

```diff
@@ -110,20 +119,78 @@ func (c *Controller) Reconcile(ctx context.Context, req reconcile.Request) (reco
 		return reconcile.Result{}, err
 	}
 
+	list := &autoscaling.ProvisioningRequestList{}
+	if err := c.client.List(ctx, list, client.InNamespace(wl.Namespace), client.MatchingFields{RequestsOwnedByWorkloadKey: wl.Name}); client.IgnoreNotFound(err) != nil {
+		return reconcile.Result{}, err
+	}
+	ownedPrs := list.Items
+	checkNameToPr := c.getCheckNameToPr(ctx, wl, relevantChecks, ownedPrs)
+
 	if workload.IsAdmitted(wl) {
 		// check the state of the provision requests, eventually toggle the checks to false
 		// otherwise there is nothing to here
 		log.V(5).Info("workload admitted, sync checks")
-		return reconcile.Result{}, c.syncCheckStates(ctx, wl, relevantChecks)
+		return reconcile.Result{}, c.syncCheckStates(ctx, wl, relevantChecks, checkNameToPr)
 	}
 
-	if err := c.syncOwnedProvisionRequest(ctx, wl, relevantChecks); err != nil {
+	err = c.deleteUnusedProvisioningRequests(ctx, ownedPrs, checkNameToPr)
+	if err != nil {
+		log.V(2).Error(err, "syncOwnedProvisionRequest failed to delete unused provisioning requests")
+		return reconcile.Result{}, err
+	}
+
+	requeAfter, err := c.syncOwnedProvisionRequest(ctx, wl, relevantChecks, checkNameToPr)
+	if err != nil {
 		// this can also delete unneeded checks
 		log.V(2).Error(err, "syncOwnedProvisionRequest failed")
 		return reconcile.Result{}, err
 	}
 
-	return reconcile.Result{}, c.syncCheckStates(ctx, wl, relevantChecks)
+	err = c.syncCheckStates(ctx, wl, relevantChecks, checkNameToPr)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	if requeAfter != nil {
+		return reconcile.Result{RequeueAfter: *requeAfter}, nil
+	}
+	return reconcile.Result{}, nil
+}
+
+func (c *Controller) getCheckNameToPr(ctx context.Context, wl *kueue.Workload, relevantChecks []string, ownedPrs []autoscaling.ProvisioningRequest) map[string]*autoscaling.ProvisioningRequest {
+	checkNameToPr := make(map[string]*autoscaling.ProvisioningRequest)
+	for _, checkName := range relevantChecks {
+		for _, pr := range ownedPrs {
+			req := &pr
+			// PRs relevant for the admission check
+			if matches(req, wl.Name, checkName) {
+				prc, err := c.helper.ProvReqConfigForAdmissionCheck(ctx, checkName)
+				if err == nil && c.reqIsNeeded(ctx, wl, prc) && requestHasParamaters(req, prc) {
+					if currPr, exists := checkNameToPr[checkName]; !exists || getAttempt(ctx, currPr, wl.Name, checkName) < getAttempt(ctx, req, wl.Name, checkName) {
```

Comment:
There should only be one ProvReq that is not Failed at a time. Furthermore, the most recent one is the one that we should care about. There shouldn't be a need to parse the number of attempts from the name.

Reply:
+1

Reply:
OTOH, the counter as part of the name has a nice implication: you can't create two ProvReq with the same name, so we don't have to worry about having more than one active ProvReq when the informer cache is slightly out of date.

Reply:
Also I would add two arguments:
- we need to keep the number of attempts somewhere, we could use annotation, but this is API change and the intention is to backport to 0.5
- it is possible that the request failed and the cloud provider needs some time to clean up nodes / resources corresponding to the old one. Starting new avoids the problem